content/style/go.md

```diff
@@ -1,3 +1,1659 @@
+    "github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+    "github.com/onflow/flow-go/model/flow"
+    "github.com/optakt/flow-dps/models/dps"
+	"sync"
+	"time"
+)
+```
+
+</td><td>
+
+```go
+import (
+    "errors"
+    "fmt"
+    "sync"
+    "time"
+
+    "github.com/rs/zerolog"
+
+    "github.com/onflow/flow-go/ledger"
+    "github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+    "github.com/onflow/flow-go/model/flow"
+
+    "github.com/optakt/flow-dps/models/dps"
+)
+```
+
+</td></tr>
+</tbody></table>
+
+### Package Names
+
+When naming packages, choose a name that is:
+
+* All lower-case. No capitals or underscores.
+* Does not need to be renamed using named imports at most call sites.
+* Short and succinct. Remember that the name is identified in full at every call site.
+* Not plural. For example, `net/url`, not `net/urls`.
+* Not "common", "util", "shared", or "lib". These are bad, uninformative names.
+
+See also [Package Names] and [Style guideline for Go packages].
+
+[Package Names]: https://blog.golang.org/package-names
+[Style guideline for Go packages]: https://rakyll.org/style-packages/
+
+### Function Names
+
+We follow the Go community's convention of using [MixedCaps for function names].
+An exception is made for test functions, which may contain underscores for the purpose of grouping related test cases, e.g., `TestMyFunction_WhatIsBeingTested`.
+
+[MixedCaps for function names]: https://golang.org/doc/effective_go.html#mixed-caps
+
+### Function Grouping and Ordering
+
+* Functions should be sorted in rough call order.
+* Functions in a file should be grouped by receiver.
+
+Therefore, exported functions should appear first in a file, after `struct`, `const`, `var` definitions.
+
+A `newXYZ()`/`NewXYZ()` may appear after the type is defined, but before the rest of the methods on the receiver.
+
+Since functions are grouped by receiver, plain utility functions should appear towards the end of the file.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+func (s *something) Cost() {
+  return calcCost(s.weights)
+}
+
+type something struct{ ... }
+
+func calcCost(n []int) int {...}
+
+func (s *something) Stop() {...}
+
+func newSomething() *something {
+    return &something{}
+}
+```
+
+</td><td>
+
+```go
+type something struct{ ... }
+
+func newSomething() *something {
+    return &something{}
+}
+
+func (s *something) Cost() {
+  return calcCost(s.weights)
+}
+
+func (s *something) Stop() {...}
+
+func calcCost(n []int) int {...}
+```
+
+</td></tr>
+</tbody></table>
+
+### Reduce Nesting
+
+Code should reduce nesting where possible by handling error cases/special conditions first and returning early or continuing the loop.
+Reduce the amount of code that is nested multiple levels.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+for _, v := range data {
+  if v.F1 == 1 {
+    v = process(v)
+    if err := v.Call(); err == nil {
```

Comment:
Something that could be mentioned are some of our established choices when writing code. Even though it makes no difference, we never write `if err := v.Call(); err == nil {`, but we always have error check in a separate line.

It can however lead to bloat and we run into the risk of documenting every single style choice we made.

Still, I think at the very least, the examples should follow our code style, and not look like they were copied from some random website.

For example, the `Exit Once` section, has this as a **good** example:

```go
func run() error {
   args := os.Args[1:]
  if len(args) != 1 {
    return errors.New("missing file")
  }
  name := args[0]
  f, err := os.Open(name)
  if err != nil {
    return err
  }
  defer f.Close()
  b, err := ioutil.ReadAll(f)
  if err != nil {
    return err
  }
  // ...
}
```

Even in this document, we emphasized the importance of adding context to errors, so it would more likely be:

```go
b, err := ioutil.ReadAll(f)
  if err != nil {
    return fmt.Errorf("could not read file: %w", err)
  }
```

Plus, code using a `b` variable would likely not pass code review 😃 
I think all examples should be demonstrative about how we really write our code.

Reply:
`b` is fine as a variable name if it's used right below.

> Variable names in Go should be short rather than long. This is especially true for local variables with limited scope. Prefer `c` to `lineCount`. Prefer `i` to `sliceIndex`.
> 
> The basic rule: the further from its declaration that a name is used, the more descriptive the name must be. For a method receiver, one or two letters is sufficient. Common variables such as loop indices and readers can be a single letter (`i`, `r`). More unusual things and global variables need more descriptive names.

Reply:
That's true, but we do have somewhat of a pattern that `io.ReadAll` or `json.Marshal` is usually `data` or `payload`

Reply:
Now that you say it, it's true that we often do that. IMO it's not great and would be better to use shorter var names when it's just used right below where it's defined and never used again. Any thoughts @ricardomgoncalves @awfm9 ?

Reply:
For this specific case, whenever I see `data` or `payload` my brain typically jumps to the conclusion that it's some binary blob. A `b` can be a binary blob, but it could also be a variable referencing a block, or a method receiver..

In general though, I don't mind something like that.

Reply:
I normally go to the data/payload var names, it is more explicit to the reader of what it is, in my opinion.

About the `if err := (..); err != nil { ... }` I actually really prefer the one line option when there is no multiple returns, the examples you presented @Maelkum are from statements with multiple return values.

Reply:
Alright, since Max isn't answering on the thread, I'll go with payload/data in the meantime so that we can get this merged. We can always change it later 🤷